### PR TITLE
WIP: Add -NoSubject switch to Send-MailMessage cmdlet

### LINKED ISSUE #4043: Send-MailMessage $Subject Mandatory=$true
Not a bug, but a request for "Send-MailMessage" cmdlet

```powershell
 $From = $email
    $To = $Number
    $Body = "Something"
    $Subject = $null
    $SMTPServer = "smtp.gmail.com"
    $SMTPPort = "587"
    Send-MailMessage -From $From -to $To -Subject $Subject `
        -Body $Body -SmtpServer $SMTPServer -port $SMTPPort -UseSsl `
    -Credential $Cred
```

RESULT:
```
Send-MailMessage : Cannot validate argument on parameter 'Subject'. The argument is null or empty. Provide an argument that is not null or empty, and then try the command again.
At line:8 char:52
+       -Body $Body -SmtpServer $SMTPServer -Subject $Subject -port $SM ...
+                                                    ~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Send-MailMessage], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.PowerShell.Commands.SendMailMessage
```

CAUSE:
```
[Parameter(Mandatory=$true, Position=1)]
        [Alias('sub')]
        [ValidateNotNullOrEmpty()]
        [string]
        ${Subject},
```

--------
Emails do not require subjects to be present.  However, in the Send-MailMessage cmdlet, the $subject parameter is set to "Mandatory=$true".  Since an email does not require a subject, this should really be set to "$false". 

I've created my own workaround this, but I feel that this should be changed since a subject is not imperative for an email to be sent.

### PR DESCRIPTION
**Pull request by ThreeFive-O**
## PR Summary

Fix #4043. 
Adds a new parameter set with `-NoSubject` switch to opt-in, which allows to send a mail message without a subject.
Adds test cases for named and positional parameters. No test case for pipeline input due to error as stated by #7591.
Adds context scopes for test cases to differentiate between parameter set tests.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link: PowerShell/PowerShell-Docs#2852
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 1]
```diff
@@ -14,7 +14,7 @@ namespace Microsoft.PowerShell.Commands
     /// <summary>
     /// Implementation for the Send-MailMessage command.
     /// </summary>
-    [Cmdlet(VerbsCommunications.Send, "MailMessage", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=135256")]
+    [Cmdlet(VerbsCommunications.Send, "MailMessage", HelpUri = "https://go.microsoft.com/fwlink/?LinkID=135256", DefaultParameterSetName="Default")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing `DefaultParameterSetName="Default"` ==> `DefaultParameterSetName = "Default"`

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 2]
```diff
@@ -58,7 +58,8 @@ public String[] Bcc
         /// <summary>
         /// Specifies the body (content) of the message.
         /// </summary>
-        [Parameter(Position = 2, ValueFromPipelineByPropertyName = true)]
+        [Parameter(Position = 2, ValueFromPipelineByPropertyName = true, ParameterSetName="Default")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 3]
```diff
@@ -58,7 +58,8 @@ public String[] Bcc
...
+        [Parameter(Position = 1, ValueFromPipelineByPropertyName = true, ParameterSetName="NoSubject")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 4]
```diff
@@ -162,7 +163,8 @@ public String From
         /// to the Powershell variable PSEmailServer, if this host can not reached an appropriate error.
         /// message will be displayed.
         /// </summary>
-        [Parameter(Position = 3, ValueFromPipelineByPropertyName = true)]
+        [Parameter(Position = 3, ValueFromPipelineByPropertyName = true, ParameterSetName="Default")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 5]
```diff
@@ -162,7 +163,8 @@ public String From
...
+        [Parameter(Position = 2, ValueFromPipelineByPropertyName = true, ParameterSetName="NoSubject")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 6]
```diff
@@ -193,7 +195,7 @@ public MailPriority Priority
         /// <summary>
         /// Specifies the subject of the email message.
         /// </summary>
-        [Parameter(Mandatory = true, Position = 1, ValueFromPipelineByPropertyName = true)]
+        [Parameter(Mandatory = true, Position = 1, ValueFromPipelineByPropertyName = true, ParameterSetName="Default")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 7]
```diff
@@ -206,6 +208,20 @@ public String Subject
         }
         private String _subject;
 
+        /// <summary>
+        /// Specifies that the mail message is sent without a subject.
+        /// </summary>
+        [Parameter(ValueFromPipelineByPropertyName = true, ParameterSetName="NoSubject")]
```
Comment (kvprasoon) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: Correct spacing too

**Review by kvprasoon — CHANGES_REQUESTED** [review]

**Comment by kvprasoon** [comment]
@ThreeFive-O Fix CodeFactor issues as well.

**Comment by ThreeFive-O** [comment]
@kvprasoon Thank you for the feedback. I've fixed the spacing issues. 

Regarding CodeFactor feedback: Most of the issues were the spacing issues as you pointed out. The other four remaining issues are styling issues which would have to apply for the entire file. But I stick to the coding style within the file. Maybe a separate PR which improves the code style would help here.

One problem I've encountered: My access to the CodeFactor feedback for this PR was denied. I had to clone the code into my own CodeFactor namespace to see the issues. Are you able to access the feedback of CodeFactor directly by clicking "Details" for this PR?

**Review by kvprasoon — APPROVED** [review]

**Comment by kvprasoon** [comment]
Nop I won't have. Cc: @iSazonov

**Comment by iSazonov** [comment]
I lost access to CodeFactor too.
/cc @TravisEz13

**Comment by TravisEz13** [comment]
I think codefactor should be back.  I had last access as well.

**Comment by iSazonov** [comment]
@ThreeFive-O You have added new switch but not implement logic to exclude a subject from the mail.

**Comment by ThreeFive-O** [comment]
@iSazonov This is intended: There is no explicit logic to leave the subject out. The implicit logic is the parameter binding. Since the named parameter `-Subject` won't be bound using the `-NoSubject` parameter, the subject of the email will be assigned to `null`. 
The possibility to get a wrong behavior would be with pipeline input, but this behavior is not correctly implemented (see also #7591), since the implemented core logic only allows attachments to be piped.
If you wish, one could implement a simple logic to set the mail subject explicitly to `null`, but due to #7591 the current implementation would have to undergo a bigger code change anyway.
In my humble opinion this PR is the simplest way to implement the behavior without changing the core logic of the code too much.

@kvprasoon  As of the CodeFactor issues: This is also intended not to be fixed in this PR, since the entire document has a different coding style. Merging this PR with all CodeFactor issues fixed would result in a mixed coding style. As cited by the coding guidelines of this repository: 

> As a general rule, our coding convention is to follow the style of the surrounding code.

**Comment by kvprasoon** [comment]
@ThreeFive-O CodeFactor checks only for changed/added lines and you have to correct it regardless of the existing code style. CodeFactor is recently added, hence not applied to existing code and I think there won't be an effort to trigger CodeFactor check against the whole codebase.

@iSazonov Correct me if I'm wrong.

**Comment by iSazonov** [comment]
@ThreeFive-O Thanks for clarify! I think it is a tricky logic and we MUST add a comment to descibe the behavior in the parameter description.

@kvprasoon We are in a transition phase and we haven't strong requirements about style issues. In the transition time we welcome and accept style fixes _in separate PRs or in separate commits_ but don't force contributes to do this.
If @ThreeFive-O have plans to continue fix issues in the cmdlet (#7591) it would be nice to fix style issues in the file beforehand.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs [hunk 8]
```diff
@@ -206,6 +208,20 @@ public String Subject
...
+        [Parameter(ValueFromPipelineByPropertyName = true, ParameterSetName = "NoSubject")]
+        public SwitchParameter NoSubject
+        {
+            get { return _nosubject; }
+            set
+            {
+                _nosubject = value;
+            }
+        }
+        private SwitchParameter _nosubject;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs]: @ThreeFive-O If you plan to continue please use auto property and remove the private field.

**Review by iSazonov — APPROVED** [review]
LGTM with two minor comments.

**Comment by iSazonov** [comment]
@dantraMSFT Could you please review the small PR?

**Comment by dantraMSFT** [comment]
>             _mMailMessage.Subject = _subject;

I would suggest setting this only if -NoSubject is not present.

Something like the following:
if (!NoSubject.IsPresent)
{
    _mMailMessage.Subject = _subject;
}

---
Refers to: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Send-MailMessage.cs:379 in 9fdce4f. [](commit_id = 9fdce4f39cae56fd5590dcf806da6bdf4dc911a8, deletion_comment = False)

**Comment by ThreeFive-O** [comment]
Implemented logic as suggested by @dantraMSFT and previously mentioned by @iSazonov.
Fixed the code factor issues as suggested by @kvprasoon.
Changed to auto property as suggested by @iSazonov.

The other style issues shall be fixed in an upcoming PR.

**Review by dantraMSFT — APPROVED** [review]
Thanks @ThreeFive-O

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this and recommends the proposal we put in the issue https://github.com/PowerShell/PowerShell/issues/4043

**Comment by iSazonov** [comment]
@ThreeFive-O Sorry, PowerShell Committee conclusion is that we should change `Subject` parameter and not add new parameter. I think you could close the PR and implement the conclusion in new PR.

**Comment by ThreeFive-O** [comment]
@iSazonov Thank you. Will make a new PR soon.

**Pull request closed without merging**